cairo/src/runner.rs

```diff
@@ -0,0 +1,472 @@
+    pub fn set_op0(&mut self) {
+        let reg = {
+            match self.instr().op0_reg() {
+                OP0_AP => self.curr.ap, // reads first word from allocated memory
+                _ => self.curr.fp,      // reads first word from input stack
+            } // no more values than 0 and 1 because op0_reg is one bit
+        };
+        self.vars.op0_addr = reg + self.instr().off_op0();
+        self.vars.op0 = self.mem.read(self.vars.op0_addr);
+    }
+
+    /// This function computes the second operand address and content and the instruction size
+    /// Panics if the flagset OP1_SRC has more than 1 nonzero bit
+    pub fn set_op1(&mut self) {
+        let (reg, size) = {
+            match self.instr().op1_src() {
+                OP1_DBL => (self.vars.op0.unwrap(), F::one()), // double indexing, op0 should be positive for address
+                OP1_VAL => (self.curr.pc, F::from(2u32)), // off_op1 will be 1 and then op1 contains an immediate value
+                OP1_FP => (self.curr.fp, F::one()),
+                OP1_AP => (self.curr.ap, F::one()),
+                _ => panic!("Invalid instruction"),
+            }
+        };
+        self.vars.size = size;
+        self.vars.op1_addr = reg + self.instr().off_op1(); // apply second offset to corresponding register
+        self.vars.op1 = self.mem.read(self.vars.op1_addr);
+    }
+
+    /// This function computes the value of the result of the arithmetic operation
+    /// Panics if a JNZ instruction is used with an invalid format
+    ///     or if the flagset RES_LOG has more than 1 nonzero bit
+    pub fn set_res(&mut self) {
+        if self.instr().pc_up() == PC_JNZ {
+            // jnz instruction
+            if self.instr().res_log() == RES_ONE
+                && self.instr().opcode() == OPC_JMP_INC
+                && self.instr().ap_up() != AP_ADD
+            {
+                self.vars.res = Some(F::zero()); // "unused"
+            } else {
+                panic!("Invalid instruction");
+            }
+        } else if self.instr().pc_up() == PC_SIZ
+            || self.instr().pc_up() == PC_ABS
+            || self.instr().pc_up() == PC_REL
+        {
+            // rest of types of updates
+            // common increase || absolute jump || relative jump
+            if self.instr().res_log() == RES_ONE {
+                self.vars.res = self.vars.op1; // right part is single operand
+            } else if self.instr().res_log() == RES_ADD {
+                self.vars.res = Some(self.vars.op0.unwrap() + self.vars.op1.unwrap());
+            // right part is addition
+            } else if self.instr().res_log() == RES_MUL {
+                self.vars.res = Some(self.vars.op0.unwrap() * self.vars.op1.unwrap());
+            // right part is multiplication
+            } else {
+                panic!("Invalid instruction");
+            }
+        } else {
+            // multiple bits take value 1
+            panic!("Invalid instruction");
```

Comment:
Actually the paper has a mixture of conditionals and match statements. But I can try to follow the paper style more literally so that I am using the exact same statement at each point.